plugins/chain_plugin/get_info_db.cpp

```diff
@@ -0,0 +1,159 @@
+
+         if (info && info->contains_full_data()) {
+            return *info;
+         }
+
+         // This only happens right after initialization when starts from
+         // snapshot as no signals are emitted. We need to cache the current states.
+         store_info();
+
+         info = std::atomic_load(&info_cache);
+         assert(info);
+         return *info;
+      }
+
+   private:
+      // A handle to the controller.
+      const chain::controller& controller;
+
+      // Indication whether get_info RPC is enabled.
+      const bool get_info_enabled = false;
+
+      // Cache to store the current get_info results.
+      // Lock free by using std::atomic_load and std::atomic_store.
+      std::shared_ptr<get_info_db::get_info_results> info_cache = nullptr;
+
+      // Fixed data
+      std::string           server_version;
+      chain::chain_id_type  chain_id;
+      std::string           server_version_string;
+      std::string           server_full_version_string;
+
+      void store_info_common(const std::shared_ptr<get_info_db::get_info_results>& info) {
+         assert(info);
+
+         // fixed part
+         info->server_version             = server_version;
+         info->chain_id                   = chain_id;
+         info->server_version_string      = server_version_string;
+         info->server_full_version_string = server_full_version_string;
+
+         // chain head part
+         const auto& head = controller.head();
+         if (head.is_valid()) {
+            info->head_block_id       = head.id();
+            info->head_block_num      = block_header::num_from_id(info->head_block_id);
+            info->head_block_time     = head.block_time();
+            info->head_block_producer = head.producer();
+         }
+
+         // fork_db part
+         if (controller.fork_db_has_root()) {
+            info->fork_db_head_block_id        = controller.fork_db_head().id();
+            info->fork_db_head_block_num       = block_header::num_from_id(*info->fork_db_head_block_id);
+            info->earliest_available_block_num = controller.earliest_available_block_num();
+         }
+
+         // resource_limits
+         const auto& rm = controller.get_resource_limits_manager();
+         info->virtual_block_cpu_limit = rm.get_virtual_block_cpu_limit();
+         info->virtual_block_net_limit = rm.get_virtual_block_net_limit();
+         info->block_cpu_limit         = rm.get_block_cpu_limit();
+         info->block_net_limit         = rm.get_block_net_limit();
+         info->total_cpu_weight        = rm.get_total_cpu_weight();
+         info->total_net_weight        = rm.get_total_net_weight();
+      }
+
+      void store_info() {
+         std::shared_ptr<get_info_db::get_info_results> info = std::make_shared<get_info_db::get_info_results>();
+
+         store_info_common(info);
+
+         if (controller.fork_db_has_root()) {
```

Comment:
Grab the `root` once like with `controller.head()` to avoid multiple mutex locks in fork db.

Reply:
https://github.com/AntelopeIO/spring/pull/920/commits/ee913fa8b0157c1dde7432c0356267fbc926abf1